# "Inferring Protein from Spectrum" solution optimization using NumPy

## Data set:

In [1]:
data_set= '''
4988.44407344
5059.48118344
5172.56524344
5328.66635344
5441.75041344
5627.82972344
5783.93083344
5940.03194344
6068.12690344
6224.22801344
6339.25495344
6442.26414344
6589.33255344
6736.40096344
6873.45987344
7029.56098344
7160.60147344
7275.62841344
7403.68699344
7540.74590344
7677.80481344
7791.84774344
7890.91615344
8076.99546344
8190.07952344
8277.11155344
8364.14358344
8421.16504344
8549.26000344
8663.30293344
8734.34004344
8791.36150344
8906.38844344
9035.43103344
9150.45797344
9263.54203344
9350.57406344
9478.63264344
9577.70105344
9705.79601344
9861.89712344
10047.9764334
10119.0135434
10275.1146534
10431.2157634
10530.2841734
10686.3852834
10789.3944734
10902.4785334
11001.5469434
11114.6310034
11229.6579434
11343.7008734
11456.7849334
11553.8376934
11656.8468834
11771.8738234
11858.9058534
11987.0008134
12044.0222734
12157.1063334
12270.1903934
12399.2329834
12530.2734734
12643.3575334
12730.3895634
12843.4736234
12900.4950834
13029.5376734
13143.5806034
13200.6020634
13314.6449934
13445.6854834
13573.7440634
13720.8124734
13857.8713834
13928.9084934
14025.9612534
14082.9827134
14170.0147434
14267.0675034
14414.1359134
14515.1835934
14618.1927834
'''

## Constants:

In [2]:
import numpy as np

AMINO_ACIDS_MASS = {
        "A": 71.03711,
        "C": 103.00919,
        "D": 115.02694,
        "E": 129.04259,
        "F": 147.06841,
        "G": 57.02146,
        "H": 137.05891,
        "K": 128.09496,
        "L": 113.08406,
        "M": 131.04049,
        "N": 114.04293,
        "P": 97.05276,
        "Q": 128.05858,
        "R": 156.10111,
        "S": 87.03203,
        "T": 101.04768,
        "V": 99.06841,
        "W": 186.07931,
        "Y": 163.06333
    }

AMINO_LABELS = np.array(list(AMINO_ACIDS_MASS.keys()))
AMINO_VALUES = np.array(list(AMINO_ACIDS_MASS.values()), dtype=np.float64)

## Solution using vanilla python:

In [3]:
def vanilla_solution(mass_string, mass_table):        
    mass_list = [float(mass.strip()) for mass in mass_string.strip().splitlines()]
    protein_string = ""
    
    for i in range(len(mass_list) - 1):
        protein_mass = abs(mass_list[i] - mass_list[i + 1])
        closest_mass_protein = ''
        for protein, mass in mass_table.items():
            if abs(protein_mass - mass) < 0.0001:
                closest_mass_protein = protein
        protein_string += closest_mass_protein
    return protein_string

- Time complexity: $O(N)$
- Space complexity: $O(N)$

In [4]:
%timeit vanilla_solution(data_set, AMINO_ACIDS_MASS)

350 μs ± 17.4 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [5]:
answer = vanilla_solution(data_set, AMINO_ACIDS_MASS)
print(answer)

ALRLWRRKRDCFFHRMDQHHNVWLSSGKNAGDEDLSQVKRWARRVRCLVLDNLPCDSKGLLEMLSLGENGNMQFHAPGSPFTC


## Solution using numpy:

In [6]:
import numpy as np

def numpy_solution(mass_string, labels_arr, values_arr):
    gen = (float(mass.strip()) for mass in mass_string.strip().splitlines())
    mass_arr = np.fromiter(gen, dtype=float)
    diffs = np.abs(np.diff(mass_arr))
    indices = np.abs(diffs[:, None] - values_arr).argmin(axis=1)
    protein_string = "".join(labels_arr[indices])
    return protein_string

- Time complexity: $O(N)$
- Space complexity: $O(N)$

In [7]:
%timeit numpy_solution(data_set, AMINO_LABELS, AMINO_VALUES)

179 μs ± 12 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [8]:
answer = numpy_solution(data_set, AMINO_LABELS, AMINO_VALUES)
print(answer)

ALRLWRRKRDCFFHRMDQHHNVWLSSGKNAGDEDLSQVKRWARRVRCLVLDNLPCDSKGLLEMLSLGENGNMQFHAPGSPFTC


**Summary:** NumPy significantly outperforms vanilla Python even on this moderately-sized dataset. The NumPy implementation completed in **179 μs**, while the vanilla solution took **350 μs** — making NumPy more than ~2x faster. This is because the vectorized broadcasting operation (`diffs[:, None] - values_arr`) eliminates the nested loop overhead inherent in the vanilla approach.